### DISCUSSION TITLE
Zig 函数调用时，传值 or 传引用

### DISCUSSION THREAD

**Discussion by jiacai2050** [zigcc/forum#69] [General]
https://www.1a-insec.net/blog/25-zig-reference-semantics/

> Primitive types such as [Integers](https://ziglang.org/documentation/0.11.0/#Integers) and [Floats](https://ziglang.org/documentation/0.11.0/#Floats) passed as parameters are copied, and then the copy is available in the function body. This is called "passing by value". Copying a primitive type is essentially free and typically involves nothing more than setting a register.

> Structs, unions, and arrays can sometimes be more efficiently passed as a reference, since a copy could be arbitrarily expensive depending on the size. When these types are passed as parameters, Zig may choose to copy and pass by value, or pass by reference, whichever way Zig decides will be faster. This is made possible, in part, by the fact that parameters are immutable.

https://ziglang.org/documentation/0.11.0/#Pass-by-value-Parameters

**Comment by jinzhongjia** [comment]
函数参数传递值的问题的话，基本类型肯定是直接复制，因为只涉及到寄存器的拷贝，但关于复杂类型的传递，并没有明确说明具体是引用还是值传递，但zig保证了参数是常量 const，这里有个相关的issue提到了一些问题，https://github.com/ziglang/zig/issues/5973

后面这个情况或许会做出一些改变

**Reply by jinzhongjia** [reply]
目前的测试是struct基本都是引用传递，但根据telegram的英文群组里讨论是该情况并不绝对

**Reply by jiacai2050** [reply]
上面的链接提到了，这由编译器自己决定。
